src/extra/recipes/grating.py

```diff
@@ -0,0 +1,172 @@
+from typing import Optional, Union, Dict, List, Tuple, Any
+from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
+
+import numpy as np
+import xarray as xr
+import h5py
+
+from extra_data import open_run, by_id, DataCollection
+from scipy.ndimage import rotate
+from extra.components import Scan
+from scipy.stats import linregress
+from functools import partial
+
+from .base import BaseCalibration
+import logging
+
+def calc_mean(energy_id: int, scan: Scan, mono_run: DataCollection,
+             grating_source: str, grating_key: str):
+    energy, train_ids = scan.steps[energy_id]
+    logging.info(f"Energy {energy}, energy id {energy_id}")
+    data = mono_run.select_trains(by_id[list(train_ids)])[grating_source, grating_key].xarray()
+    return data.mean('trainId').to_numpy()
+
+class Grating2DCalibration(BaseCalibration):
+    """
+    Calibrate a 2D grating spectrometer.
+
+    Args:
+      grating_source: Where to read the grating data from.
+      angle: The rotation angle in degrees if the camera is not aligned.
+      energy_source: Where to read the undulator energy from.
+    """
+    def __init__(self,
+                 grating_source: str='SQS_DIAG3_BIU/CAM/CAM_6:daqOutput',
```

Comment:
Is there only one of these at XFEL?

Reply:
You are right ... I moved them to `setup()` and removed the default argument.

Reply:
Well, if there is only one of them for the foreseeable future then I don't think there's anything wrong with leaving them as default arguments :) But if more exist/are expected then it might be worth trying to automatically find them based on the device class or something, like the XGM and digitizer components.